### PR TITLE
[sub intf] Fix kernel side processing to enslave sub interface to non-default vrf

### PR DESCRIPTION
**Pull request by wendani**
**What I did**
Change in IntfMgr:
1) Use `alias` (instead of `subIntfAlias` before the change)  to hold complete sub interface name; use `parentAlias` (instead of `alias` before the change) to hold parent port name. 
2) Move sub interface creation to the first place before processing fields vrf, proxy arp, and garp.

In doing above, a sub interface can receive the same treatment as other types of interfaces (interface type port, type vlan) on vrf, proxy arp, and garp fields, while maintaining an indicator (parentAlias not empty) to receive treatment specific to a sub interface (creation, deletion).

**Why I did it**
Fix https://github.com/Azure/sonic-swss/issues/1277, https://github.com/Azure/sonic-swss/issues/1510

**How I verified it**
vs test. Extended all test cases to test a sub interface ingress linked to a non-default vrf.

Before the change the following extended test cases fail:

1. `test_sub_port_intf_creation`
```
========================================================================== FAILURES ==========================================================================
________________________________________________________ TestSubPortIntf.test_sub_port_intf_creation _________________________________________________________

self = <test_sub_port_intf.TestSubPortIntf object at 0x7fb8e66d92e8>, dvs = <conftest.DockerVirtualSwitch object at 0x7fb8ea94f5f8>

    def test_sub_port_intf_creation(self, dvs):
        self.connect_dbs(dvs)
    
        self._test_sub_port_intf_creation(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST)
        self._test_sub_port_intf_creation(dvs, self.LAG_SUB_PORT_INTERFACE_UNDER_TEST)
    
>       self._test_sub_port_intf_creation(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST, self.VRF_UNDER_TEST)

test_sub_port_intf.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_sub_port_intf.py:267: in _test_sub_port_intf_creation
    self.check_sub_port_intf_vrf_bind_kernel(dvs, sub_port_intf_name, vrf_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test_sub_port_intf.TestSubPortIntf object at 0x7fb8e66d92e8>, dvs = <conftest.DockerVirtualSwitch object at 0x7fb8ea94f5f8>
port_name = 'Ethernet64.10', vrf_name = 'Vrf0'

    def check_sub_port_intf_vrf_bind_kernel(self, dvs, port_name, vrf_name):
        (ec, out) = dvs.runcmd(['bash', '-c', "ip link show {} | grep {}".format(port_name, vrf_name)])
>       assert ec == 0
E       assert 1 == 0
E         +1
E         -0

test_sub_port_intf.py:207: AssertionError
================================================================== short test summary info ===================================================================
FAILED test_sub_port_intf.py::TestSubPortIntf::test_sub_port_intf_creation - assert 1 == 0
================================================================ 1 failed in 60.83s (0:01:00) ================================================================
```

2. `test_sub_port_intf_add_ip_addrs`
```
========================================================================== FAILURES ==========================================================================
______________________________________________________ TestSubPortIntf.test_sub_port_intf_add_ip_addrs _______________________________________________________

self = <test_sub_port_intf.TestSubPortIntf object at 0x7f91ee6166a0>, dvs = <conftest.DockerVirtualSwitch object at 0x7f91ee633518>

    def test_sub_port_intf_add_ip_addrs(self, dvs):
        self.connect_dbs(dvs)
    
        self._test_sub_port_intf_add_ip_addrs(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST)
        self._test_sub_port_intf_add_ip_addrs(dvs, self.LAG_SUB_PORT_INTERFACE_UNDER_TEST)
    
>       self._test_sub_port_intf_add_ip_addrs(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST, self.VRF_UNDER_TEST)

test_sub_port_intf.py:380: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_sub_port_intf.py:356: in _test_sub_port_intf_add_ip_addrs
    self.check_sub_port_intf_route_entries(vrf_oid)
test_sub_port_intf.py:203: in check_sub_port_intf_route_entries
    wait_for_result(_access_function, DVSDatabase.DEFAULT_POLLING_CONFIG)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

polling_function = <function TestSubPortIntf.check_sub_port_intf_route_entries.<locals>._access_function at 0x7f91ee688730>
polling_config = PollingConfig(polling_interval=0.01, timeout=5, strict=True)

    def wait_for_result(
        polling_function: Callable[[], Tuple[bool, Any]],
        polling_config: PollingConfig,
    ) -> Tuple[bool, Any]:
        """Run `polling_function` periodically using the specified `polling_config`.
    
        Args:
            polling_function: The function being polled. The function cannot take any arguments and
                must return a status which indicates if the function was succesful or not, as well as
                some return value.
            polling_config: The parameters to use to poll the polling function.
    
        Returns:
            If the polling function succeeds, then this method will return True and the output of the
            polling function.
    
            If it does not succeed within the provided timeout, it will return False and whatever the
            output of the polling function was on the final attempt.
        """
        if polling_config.polling_interval == 0:
            iterations = 1
        else:
            iterations = int(polling_config.timeout // polling_config.polling_interval) + 1
    
        for _ in range(iterations):
            status, result = polling_function()
    
            if status:
                return (True, result)
    
            time.sleep(polling_config.polling_interval)
    
        if polling_config.strict:
>           assert False, f"Operation timed out after {polling_config.timeout} seconds"
E           AssertionError: Operation timed out after 5 seconds

dvslib/dvs_common.py:54: AssertionError
================================================================== short test summary info ===================================================================
FAILED test_sub_port_intf.py::TestSubPortIntf::test_sub_port_intf_add_ip_addrs - AssertionError: Operation timed out after 5 seconds
================================================================ 1 failed in 69.03s (0:01:09) ================================================================
```

3. `test_sub_port_intf_removal`
```
========================================================================== FAILURES ==========================================================================
_________________________________________________________ TestSubPortIntf.test_sub_port_intf_removal _________________________________________________________

self = <test_sub_port_intf.TestSubPortIntf object at 0x7fe9cc3aa438>, dvs = <conftest.DockerVirtualSwitch object at 0x7fe9cc3aa518>

    def test_sub_port_intf_removal(self, dvs):
        self.connect_dbs(dvs)
    
        self._test_sub_port_intf_removal(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST)
        self._test_sub_port_intf_removal(dvs, self.LAG_SUB_PORT_INTERFACE_UNDER_TEST)
    
>       self._test_sub_port_intf_removal(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST, self.VRF_UNDER_TEST)

test_sub_port_intf.py:635: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
test_sub_port_intf.py:580: in _test_sub_port_intf_removal
    self.check_sub_port_intf_vrf_bind_kernel(dvs, sub_port_intf_name, vrf_name)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <test_sub_port_intf.TestSubPortIntf object at 0x7fe9cc3aa438>, dvs = <conftest.DockerVirtualSwitch object at 0x7fe9cc3aa518>
port_name = 'Ethernet64.10', vrf_name = 'Vrf0'

    def check_sub_port_intf_vrf_bind_kernel(self, dvs, port_name, vrf_name):
        (ec, out) = dvs.runcmd(['bash', '-c', "ip link show {} | grep {}".format(port_name, vrf_name)])
>       assert ec == 0
E       assert 1 == 0
E         +1
E         -0

test_sub_port_intf.py:207: AssertionError
================================================================== short test summary info ===================================================================
FAILED test_sub_port_intf.py::TestSubPortIntf::test_sub_port_intf_removal - assert 1 == 0
================================================================ 1 failed in 61.08s (0:01:01) ================================================================
```

4. removal sequence test, piggy-backed onto `test_sub_port_intf_removal`

Validate that sub interface removal does not take effect until all IP addresses configured on it are removed.

Before the change, removal sequence test fails:
```
=========================================================================== FAILURES ============================================================================
__________________________________________________________ TestSubPortIntf.test_sub_port_intf_removal ___________________________________________________________

self = <test_sub_port_intf.TestSubPortIntf object at 0x7fbb75b9fb38>, dvs = <conftest.DockerVirtualSwitch object at 0x7fbb75b9f7b8>

    def test_sub_port_intf_removal(self, dvs):
        self.connect_dbs(dvs)
    
        #self._test_sub_port_intf_removal(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST)
        #self._test_sub_port_intf_removal(dvs, self.LAG_SUB_PORT_INTERFACE_UNDER_TEST)
    
        #self._test_sub_port_intf_removal(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST, vrf_name=self.VRF_UNDER_TEST)
        #self._test_sub_port_intf_removal(dvs, self.LAG_SUB_PORT_INTERFACE_UNDER_TEST, vrf_name=self.VRF_UNDER_TEST)
    
        #self._test_sub_port_intf_removal(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST, vrf_name=self.VNET_UNDER_TEST)
        #self._test_sub_port_intf_removal(dvs, self.LAG_SUB_PORT_INTERFACE_UNDER_TEST, vrf_name=self.VNET_UNDER_TEST)
    
>       self._test_sub_port_intf_removal(dvs, self.SUB_PORT_INTERFACE_UNDER_TEST, removal_seq_test=True)

test_sub_port_intf.py:950: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
test_sub_port_intf.py:851: in _test_sub_port_intf_removal
    self.check_sub_port_intf_fvs(self.state_db, state_tbl_name, sub_port_intf_name, fv_dict)
test_sub_port_intf.py:290: in check_sub_port_intf_fvs
    db.wait_for_field_match(table_name, key, fv_dict)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <dvslib.dvs_database.DVSDatabase object at 0x7fbb75bea8d0>, table_name = 'PORT_TABLE', key = 'Ethernet64.10', expected_fields = {'state': 'ok'}
polling_config = PollingConfig(polling_interval=0.01, timeout=5.0, strict=True), failure_message = None

    def wait_for_field_match(
        self,
        table_name: str,
        key: str,
        expected_fields: Dict[str, str],
        polling_config: PollingConfig = PollingConfig(),
        failure_message: str = None,
    ) -> Dict[str, str]:
        """Wait for the entry stored at `key` to have the specified field/values and retrieve it.
    
        This method is useful if you only care about the contents of a subset of the fields stored
        in the specified entry.
    
        Args:
            table_name: The name of the table where the entry is stored.
            key: The key that maps to the entry being checked.
            expected_fields: The fields and their values we expect to see in the entry.
            polling_config: The parameters to use to poll the db.
            failure_message: The message to print if the call times out. This will only take effect
                if the PollingConfig is set to strict.
    
        Returns:
            The entry stored at `key`. If no entry is found, then an empty Dict is returned.
        """
    
        def access_function():
            fv_pairs = self.get_entry(table_name, key)
            return (
                all(fv_pairs.get(k) == v for k, v in expected_fields.items()),
                fv_pairs,
            )
    
        status, result = wait_for_result(
            access_function, self._disable_strict_polling(polling_config)
        )
    
        if not status:
            message = failure_message or (
                f"Expected field/value pairs not found: expected={expected_fields}, "
                f'received={result}, key="{key}", table="{table_name}"'
            )
>           assert not polling_config.strict, message
E           AssertionError: Expected field/value pairs not found: expected={'state': 'ok'}, received={}, key="Ethernet64.10", table="PORT_TABLE"

dvslib/dvs_database.py:203: AssertionError
==================================================================== short test summary info ====================================================================
FAILED test_sub_port_intf.py::TestSubPortIntf::test_sub_port_intf_removal - AssertionError: Expected field/value pairs not found: expected={'state': 'ok'}, re...
================================================================= 1 failed in 69.54s (0:01:09) ==================================================================
```

**Details if related**

#### Dependancy

- [x] Contains https://github.com/Azure/sonic-swss/pull/1654

### CODE REVIEW

**Comment by wendani** [comment]
retest vs please

File: cfgmgr/intfmgr.cpp [hunk 1]
```diff
@@ -576,10 +529,57 @@ bool IntfMgr::doIntfGeneralTask(const vector<string>& keys,
...
+            }
+
+            if (!alias.compare(0, strlen(VLAN_PREFIX), VLAN_PREFIX))
+            {
+                FieldValueTuple fvTuple("grat_arp", grat_arp);
+                data.push_back(fvTuple);
+            }
+        }
+
+        m_appIntfTableProducer.set(alias, data);
```
Comment (prsunny) [inline cfgmgr/intfmgr.cpp:581]: Does orchagent requires similar change to handle sub-interface RIF configurations?

Comment (wendani) [reply cfgmgr/intfmgr.cpp:581]: No, ASIC_DB output looks good.

**Comment by wendani** [comment]
retest vs please

**Comment by wendani** [comment]
retest vs please

**Comment by wendani** [comment]
retest vs please

**Comment by wendani** [comment]
retest Azure.sonic-swss please

**Comment by liat-grozovik** [comment]
@prsunny  should we have this fixed on 202012 as well? if so, can you please add label?

**Comment by prsunny** [comment]
@liat-grozovik , it seems to have significant changes and if we are not blocked on this for 202012, prefer to have this on master only.

**Comment by liat-grozovik** [comment]
NP

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________

**Review by prsunny — APPROVED** [review]

**Review by neethajohn — APPROVED** [review]

**Comment by neethajohn** [comment]
Discussed with Prince offline on taking it to 202012 branch

**Pull request merged**